### PR TITLE
Fix #9706 PathLike with subtype

### LINKED ISSUE #9706: PydanticSchemaGenerationError when using os.PathLike[str]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Works with plain `os.PathLike`. Error with `os.PathLike[str]`, `os.PathLike[bytes]`, `os.PathLike[Any]`.

```
$ python main.py 
Traceback (most recent call last):
  File "main.py", line 5, in <module>
    class Model(pydantic.BaseModel):
  File "pydantic/_internal/_model_construction.py", line 202, in __new__
    complete_model_class(
  File "pydantic/_internal/_model_construction.py", line 539, in complete_model_class
    schema = cls.__get_pydantic_core_schema__(cls, handler)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pydantic/main.py", line 626, in __get_pydantic_core_schema__
    return handler(source)
           ^^^^^^^^^^^^^^^
  File "pydantic/_internal/_schema_generation_shared.py", line 82, in __call__
    schema = self._handler(source_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pydantic/_internal/_generate_schema.py", line 502, in generate_schema
    schema = self._generate_schema_inner(obj)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pydantic/_internal/_generate_schema.py", line 753, in _generate_schema_inner
    return self._model_schema(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "pydantic/_internal/_generate_schema.py", line 580, in _model_schema
    {k: self._generate_md_field_schema(k, v, decorators) for k, v in fields.items()},
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pydantic/_internal/_generate_schema.py", line 580, in <dictcomp>
    {k: self._generate_md_field_schema(k, v, decorators) for k, v in fields.items()},
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pydantic/_internal/_generate_schema.py", line 916, in _generate_md_field_schema
    common_field = self._common_field_schema(name, field_info, decorators)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pydantic/_internal/_generate_schema.py", line 1081, in _common_field_schema
    schema = self._apply_annotations(
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "pydantic/_internal/_generate_schema.py", line 1825, in _apply_annotations
    schema = get_inner_schema(source_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pydantic/_internal/_schema_generation_shared.py", line 82, in __call__
    schema = self._handler(source_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pydantic/_internal/_generate_schema.py", line 1806, in inner_handler
    schema = self._generate_schema_inner(obj)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pydantic/_internal/_generate_schema.py", line 758, in _generate_schema_inner
    return self.match_type(obj)
           ^^^^^^^^^^^^^^^^^^^^
  File "pydantic/_internal/_generate_schema.py", line 840, in match_type
    return self._match_generic_type(obj, origin)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pydantic/_internal/_generate_schema.py", line 888, in _match_generic_type
    return self._unknown_type_schema(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pydantic/_internal/_generate_schema.py", line 405, in _unknown_type_schema
    raise PydanticSchemaGenerationError(
pydantic.errors.PydanticSchemaGenerationError: Unable to generate pydantic-core schema for os.PathLike[str]. Set `arbitrary_types_allowed=True` in the model_config to ignore this error or implement `__get_pydantic_core_schema__` on your type to fully support it.

If you got this error by calling handler(<some type>) within `__get_pydantic_core_schema__` then you likely need to call `handler.generate_schema(<some type>)` since we do not call `__get_pydantic_core_schema__` on `<some type>` otherwise to avoid infinite recursion.

For further information visit https://errors.pydantic.dev/2.7/u/schema-for-unknown-type
```

### Example Code

```Python
# main.py

import os
import pydantic

class Model(pydantic.BaseModel):
    field: os.PathLike[str]
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.7.3
        pydantic-core version: 2.18.4
          pydantic-core build: profile=release pgo=true
                 install path: pydantic
               python version: 3.11.9 (main, Apr  6 2024, 17:59:24) [GCC 11.4.0]
                     platform: Linux-6.5.0-35-generic-x86_64-with-glibc2.35
             related packages: typing_extensions-4.12.2
                       commit: unknown
```

### PR DESCRIPTION
**Pull request by nix010**
## Change Summary

Fix #9706 

## Related issue number

Fix #9706 

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @hramezani

### CODE REVIEW

**Comment by nix010** [comment]
please review

**Comment by nix010** [comment]
Not sure if the change is align wiht what is expected from #9706 but made PR anyway to see if it on the right direction!

File: pydantic/_internal/_std_types_schema.py [hunk 1]
```diff
@@ -208,6 +208,9 @@ def path_schema_prepare_pydantic_annotations(
     import pathlib
 
     if source_type not in {
+        os.PathLike[bytes],
+        os.PathLike[str],
+        os.PathLike[Any],
```
Comment (sydney-runkle) [inline pydantic/_internal/_std_types_schema.py:213]: I'd rather us use `get_origin` on `source_type`, and if that's `os.PathLike`, then we can move forward with building a validator that takes into account how `os.PathLike` is parametrized

File: tests/test_types.py [hunk 2]
```diff
@@ -3494,6 +3494,32 @@ class Model(BaseModel):
     }
 
 
+def test_path_like_extra_subtype():
```
Comment (sydney-runkle) [inline tests/test_types.py:3497]: Test looks great, thanks!

Could you please add some tests that verify validation errors are thrown when invalid data is passed in for each of these parametrized types?

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
Hey @nix010,

Great start here. Left some general feedback + a few test requests!

File: tests/test_types.py [hunk 3]
```diff
@@ -3494,6 +3494,44 @@ class Model(BaseModel):
...
+    with pytest.raises(ValidationError) as exc_info:
+        Model(
+            str_type=b'/foo/bar',
+            byte_type='/foo/bar',
+            any_type=111,
+        )
+    assert exc_info.value.errors(include_url=False) == [
+        {'type': 'path_type', 'loc': ('str_type',), 'msg': 'Input is not a valid path', 'input': b'/foo/bar'},
+        {'type': 'path_type', 'loc': ('byte_type',), 'msg': 'Input is not a valid path', 'input': '/foo/bar'},
+        {'type': 'path_type', 'loc': ('any_type',), 'msg': 'Input is not a valid path', 'input': 111},
```
Comment (nix010) [inline tests/test_types.py:3531]: I tried to change the msg to something more meaningful than `Input is not a valid path` but had some trouble with `lax_schema`
@sydney-runkle any suggestions ?
https://github.com/pydantic/pydantic/blob/f024d03b832d1bbcbadf76184ed14d92571a71ca/pydantic/_internal/_std_types_schema.py#L244-L252

File: pydantic/_internal/_std_types_schema.py [hunk 4]
```diff
@@ -207,7 +207,11 @@ def path_schema_prepare_pydantic_annotations(
 ) -> tuple[Any, list[Any]] | None:
     import pathlib
 
-    if source_type not in {
+    orig_source_type = get_origin(source_type) or source_type
```
Comment (sydney-runkle) [inline pydantic/_internal/_std_types_schema.py:210]: What about:

```py
orig_source_type: Any = get_origin(source_type) or source_type
if (source_type_args := get_args(source_type)) and source_type_args[0] not in {str, bytes, Any}:
    return None

if orig_source_type not in {
    os.PathLike,
    pathlib.Path,
    pathlib.PurePath,
    pathlib.PosixPath,
    pathlib.PurePosixPath,
    pathlib.PureWindowsPath,
}:
    return None
```

File: pydantic/_internal/_std_types_schema.py [hunk 5]
```diff
@@ -217,22 +221,46 @@ def path_schema_prepare_pydantic_annotations(
...
+    )
+
+    if is_path_like_subtype_invalid:
+        return None
+
     metadata, remaining_annotations = _known_annotated_metadata.collect_known_metadata(annotations)
-    _known_annotated_metadata.check_metadata(metadata, _known_annotated_metadata.STR_CONSTRAINTS, source_type)
+    _known_annotated_metadata.check_metadata(metadata, _known_annotated_metadata.STR_CONSTRAINTS, orig_source_type)
+
+    construct_path = orig_source_type
```
Comment (sydney-runkle) [inline pydantic/_internal/_std_types_schema.py:241]: What about:

```py
construct_path = pathlib.PurePath if orig_source_type is os.PathLike else orig_source_type
constrained_schema = (
    core_schema.bytes_schema(**metadata) if source_type_args and source_type_args[0] is bytes 
    else core_schema.str_schema(**metadata)
)

def path_validator(input_value: str) -> os.PathLike[Any]:
    try:
        return construct_path(input_value)
    except TypeError as e:
        raise PydanticCustomError('path_type', 'Input is not a valid path') from e
```

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
Some ideas for simplification :)

File: tests/test_types.py [hunk 6]
```diff
@@ -3494,6 +3494,44 @@ class Model(BaseModel):
...
+        byte_type=b'/foo/bar',
+        any_type='/foo/bar',
+    )
+    assert m.str_type == Path('/foo/bar')
+    assert m.byte_type == Path('/foo/bar')
+    assert m.any_type == Path('/foo/bar')
+    assert Model.model_json_schema() == {
+        'properties': {
+            'str_type': {'format': 'path', 'title': 'Str Type', 'type': 'string'},
+            'byte_type': {'format': 'path', 'title': 'Byte Type', 'type': 'string'},
```
Comment (sydney-runkle) [inline tests/test_types.py:3514]: Hmm, I wonder if we want to change the JSON schema for these path types...

File: pydantic/_internal/_std_types_schema.py (hunk 5 above)
Comment (sydney-runkle) [reply pydantic/_internal/_std_types_schema.py:241]: I realize I've skipped the decoding here, one moment... trying to add back in with correct typing

Comment (sydney-runkle) [reply pydantic/_internal/_std_types_schema.py:241]: I think this is type safe:

```py
def path_validator(input_value: str | bytes) -> os.PathLike[str]:
    try:
        if source_type_args and source_type_args[0] is bytes:
            if isinstance(input_value, bytes):
                try:
                    input_value = input_value.decode()
                except UnicodeDecodeError as e:
                    raise PydanticCustomError('bytes_type', 'Input must be valid bytes') from e
            else:
                raise PydanticCustomError('bytes_type', 'Input must be bytes')
        elif not isinstance(input_value, str):
            raise PydanticCustomError('path_type', 'Input is not a valid path')
    
        return construct_path(input_value)
    except TypeError as e:
        raise PydanticCustomError('path_type', 'Input is not a valid path') from e
```

File: tests/test_types.py (hunk 3 above)
Comment (sydney-runkle) [reply tests/test_types.py:3531]: Yeah the difficulty here is that if you don't override the errors from the union schema, you get kind of a nasty spiral bc of the checks against multiple members

Comment (sydney-runkle) [reply tests/test_types.py:3531]: Why don't we just customize it like this:

```py
schema = core_schema.lax_or_strict_schema(
        lax_schema=core_schema.union_schema(
            [
                instance_schema,
                core_schema.no_info_after_validator_function(path_validator, constrained_schema),
            ],
            custom_error_type='path_type',
            custom_error_message=f'Input is not a valid path for {source_type}',
            strict=True,
        ),
        strict_schema=instance_schema,
        serialization=core_schema.to_string_ser_schema(),
        strict=strict,
    )
```

Then we can make the error handling more simple in `path_validator`

File: tests/test_types.py (hunk 6 above)
Comment (nix010) [reply tests/test_types.py:3514]: sure, what should we change them into ?

File: pydantic/_internal/_std_types_schema.py (hunk 4 above)
Comment (sydney-runkle) [reply pydantic/_internal/_std_types_schema.py]: @nix010,

I believe my suggestion above does the same...

File: tests/test_types.py (hunk 6 above)
Comment (sydney-runkle) [reply tests/test_types.py:3514]: I feel like the `type` should be `bytes` for the `byte_type` path...

Comment (sydney-runkle) [reply tests/test_types.py:3514]: Honestly though, if someone complains we can fix this, I don't think it's super high priority and this gives us some flexibility to change it based on user experience.

Comment (sydney-runkle) [reply tests/test_types.py:3514]: Bc it's really a string once you populate the path

**Review by sydney-runkle — APPROVED** [review]
Great, thanks!!

**Pull request merged**